Add async support

stevefan1999-personal opened the pull request.

stevefan1999-personal committed:
add simd support

stevefan1999-personal committed:
add async feature

stevefan1999-personal committed:
remove anyhow requirement

stevefan1999-personal committed:
reformat example server

ninjasource commented on examples/server/main.rs:
```diff
@@ -95,85 +108,65 @@ fn handle_client(mut stream: TcpStream) -> Result<()> {
                 WebSocketSendMessageType::Text,
                 true,
-                text.as_bytes(),
+                format!("shut the fuck up {}", text).as_bytes(),
```
What's this?

stevefan1999-personal replied:
oh wait

stevefan1999-personal replied:
just uhhhhhhhhh let's say it never happened

ninjasource commented on Cargo.toml:
```diff
@@ -11,17 +11,69 @@ categories = ["embedded", "no-std", "network-programming"]
+rand_core = "0.6.3"
+sha1 = "0.10.1"
+smol = { version = "1.2.5", optional = true }
+tokio = { version = "1.19.2", features = ["io-util"], optional = true }
```
At this point in time I do not wish to add tokio, smol, async-std and all their baggage to this codebase. I do plan on adding async support but I wanted to think about it properly before attempting it. I think the async streaming story in Rust is a tire fire and I don't want to add to it.

stevefan1999-personal replied:
ok, we should keep async feature then

stevefan1999-personal replied:
and then for each stream the user will need to provide their own compat implementation

ninjasource replied:
Yes I think that would be better

stevefan1999-personal committed:
move Write and Read definition to its own compat crate

since these two types are mimicking std extensions, it is better to denote them as compat

stevefan1999-personal committed:
remove async framework dependencies, add example compat converter to each async example

stevefan1999-personal committed:
remove core2 dependencies

stevefan1999-personal committed:
Merge remote-tracking branch 'upstream/master' into patch-async

# Conflicts:
#	Cargo.toml

stevefan1999-personal committed:
return error early for other httparse error rather than panicking

stevefan1999-personal commented:
hmm, speaking of the `Read` and `Write` support, [embedded_io](https://docs.rs/embedded-io/latest/embedded_io/) should actually worth to take a look. Even if it looks like there aren't `AsyncRead` and `AsyncWrite` yet

elpiel commented:
Any progress on this async support @ninjasource ?

ninjasource commented:
Thanks @elpiel , I'll take a loot at async again. Been working on other stuff.

elpiel commented:
I'm open in helping out with this effort as well. Can't wait to try out embedded websockets!

elpiel commented:
embedded-hal and embedded-hal-async are now on version 1.0 :tada:
https://docs.rs/embedded-hal-async/
